lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php

```diff
@@ -87,4 +87,132 @@ protected static function get_blocks_metadata() {
 
 		return static::$blocks_metadata;
 	}
+
+	/**
+	 * Builds metadata for the style nodes, which returns in the form of:
+	 *
+	 *     [
+	 *       [
+	 *         'path'     => [ 'path', 'to', 'some', 'node' ],
+	 *         'selector' => 'CSS selector for some node',
+	 *         'duotone'  => 'CSS selector for duotone for some node'
+	 *       ],
+	 *       [
+	 *         'path'     => ['path', 'to', 'other', 'node' ],
+	 *         'selector' => 'CSS selector for other node',
+	 *         'duotone'  => null
+	 *       ],
+	 *     ]
+	 *
+	 * @since 5.8.0
+	 *
+	 * @param array $theme_json     The tree to extract style nodes from.
+	 * @return array
+	 */
+	protected static function get_style_nodes( $theme_json ) {
+		$nodes = array();
+		if ( ! isset( $theme_json['styles'] ) ) {
+			return $nodes;
+		}
+
+		// Top-level.
+		$nodes[] = array(
+			'path'     => array( 'styles' ),
+			'selector' => static::ROOT_BLOCK_SELECTOR,
+		);
+
+		if ( isset( $theme_json['styles']['elements'] ) ) {
+			foreach ( $theme_json['styles']['elements'] as $element => $node ) {
+				$nodes[] = array(
+					'path'     => array( 'styles', 'elements', $element ),
+					'selector' => static::ELEMENTS[ $element ],
```

Comment:
In unrelated testing in Core, I'm seeing notices coming from here:

```
Notice: Undefined index: button in /var/www/html/wp-includes/class-wp-theme-json.php on line 1482
```

@scruffian: can someone follow up here to not assume `$element` exists on the table? Otherwise this will happen whenever someone is using a recent theme with a less recent Core installation, for example.

Reply:
I've run into similar things a few times ([here](https://github.com/WordPress/gutenberg/pull/39529#issuecomment-1154709502) and [here](https://github.com/WordPress/gutenberg/pull/43526#issuecomment-1226703539)). 

In Core, with Gutenberg activated, [this line calls get_block_editor_settings()](https://github.com/WordPress/wordpress-develop/blob/b316c8b25fc71920d89ee37ee26609a77b2a305a/src/wp-admin/edit-form-blocks.php#L282) which calls `wp_get_global_stylesheet()`, and the subsequent chain of methods.


```
<b>Notice</b>:  Undefined index: button in <b>/var/www/src/wp-includes/class-wp-theme-json.php</b> on line <b>1489</b><br />
#0  WP_Theme_JSON::get_style_nodes() called at [/var/www/src/wp-includes/class-wp-theme-json.php:712]
#1  WP_Theme_JSON->get_stylesheet() called at [/var/www/src/wp-includes/global-styles-and-settings.php:140]
#2  wp_get_global_stylesheet()) called at [/var/www/src/wp-includes/block-editor.php:404]
#3  get_block_editor_settings() called at [/var/www/src/wp-admin/post.php:187]
```

So it looks like it's running the Core methods/classes which may not have the latest checks ?

I've added a check in Gutenberg:

- https://github.com/WordPress/gutenberg/pull/43622

I tried to find an "in-plugin" work around and discovered the following:

Removing the call to `WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()` in [gutenberg_register_webfonts_from_theme_json](https://github.com/WordPress/gutenberg/blob/877668d19f934d39b25ac351751620207f727b18/lib/experimental/register-webfonts-from-theme-json.php#L13) removes the error. I don't know why 🤷 

The WP_Theme_JSON constructor is called twice where [static:ELEMENTS](https://github.com/ramonjd/wordpress-develop/blob/c4d2ffd4b4383a3ef319b91ebea587336de35d1e/src/wp-includes/class-wp-theme-json.php#L405) refers to the [Core const](https://github.com/ramonjd/wordpress-develop/blob/c4d2ffd4b4383a3ef319b91ebea587336de35d1e/src/wp-includes/class-wp-theme-json.php#L352), that is, **without** the new elements `button` et. al., from the WP_Theme_JSON_Gutenberg class.

This is why we see the error where theme.json contains unsupported elements.

If, in the constructor, we referred to `self::ELEMENTS` we'd avoid this error, but it'd also mean we'd have to overwrite the constructor every time ELEMENTS was updated.

Reply:
> If, in the constructor, we referred to self::ELEMENTS we'd avoid this error, but it'd also mean we'd have to overwrite the constructor every time ELEMENTS was updated.

This looks like the way to go. Some copy pasta is a small price to pay for better consistency of how things work!